abi.md

```diff
@@ -0,0 +1,147 @@
+behaviors and boundary constructs in FIRRTL to System Verilog must be defined.
+Where possible this ABI does not impose constraints on implementation,
+concerning itself primarily with the boundaries of the circuit.
+
+## On Modules
+
+### The Circuit and Top Level
+
+The top level module, as specified by the circuit name, shall be present as a
+System Verilog module of the same name.  This firrtl module is considered a
+"public" module and subject to the lowering constraints for public modules.
+
+### External Modules
+
+An external module may be presumed to exist following the lowering constraints
+for public modules.  The module shall exist with a verilog name matching the 
+defname value or, lacking that, the module name.
+
+###  Public Modules
+
+Any module considered a "public" module shall be implemented in Verilog in a
+consistent way.  Any public module shall exist as a Verilog module of the same
+name.
+
+Each public module with definitions (e.g. not external modules) shall be placed
+in a file with the same name.  No assumption is made of the filename of an
+implementation of an external module, it is the user's job to include files in
+such a way in their tools to resolve the name.
+
+### Port Lowering ABIv1
+
+Ports are generally lowered to netlist types, except where Verilog's type system
+prevents it.
+
+Ports of integer types shall be lowered to netlist ports (`wire`{.verilog}) as a
+packed vector of equivalent size.  For example, consider the following FIRRTL:
+
+```firrtl
+circuit Top :
+  module Top :
+    output out: UInt<16>
+    input b: UInt<32>
+```
+
+This is lowered to the following Verilog:
+
+```verilog
+module Top(
+    output wire [15:0] out,
+    input wire [31:0] in
+);
+```
+
+Ports of aggregate type shall be lowered according to the "Aggregate Type
+Lowering" description in the firrtl spec.
+
+Ports of enum type shall be lowered to ports of a type consistent with the rules
+for lowering firrtl enum types to Verilog.
+
+Ports of ref type shall be lowered to a Verilog macro of the form `` `define
+ref_<circuit name>_<module name>_<portname> <internal path from module>`` in a
+file with name `ref_<circuit name>_<module name>.sv`.
+
+### Port Lowering ABIv2
+
+Ports are lowered per the v1 ABI above, except for aggregate types.
+
+Vectors shall be lowered to Verilog packed vectors.
+
+Bundles shall be recursively split as per "Aggregate Type Lowering", except
+instead of recursively converting bundles to ground types, the recursion stops
+at passive types.
+
+Passive bundles shall be lowered to Verilog packed structs.
+
+Reference types in ports shall be logically split out from aggregates and named
+as though "Aggregate Type Lowering" was used.
+
+## On Types
+
+Types are only guaranteed to follow this lowering when the Verilog type is
+visible for some reason related to the visibility of the element.  These include
+use in ports and elements exported by reference through a public module.
```

Comment:
This might be made clearer by defining a notion of a "public" type analogous to a public module. It isn't unclear as it stands, but it might be good to have that definition in our back pocket for later.

Reply:
Since types in firrtl are structural, I don't want to claim a particular lowering for a type, but rather for circuit elements which are visible and have that type.